src/common/oclengine.cpp

```diff
@@ -52,12 +77,97 @@ void OCLEngine::SetDeviceContextPtrVector(std::vector<DeviceContextPtr> vec, Dev
 
 void OCLEngine::SetDefaultDeviceContext(DeviceContextPtr dcp) { default_device_context = dcp; }
 
-OCLEngine::OCLEngine() { InitOCL(); }
+OCLEngine::OCLEngine()
+{
+    OCLInitResult res = InitOCL(false);
+    all_device_contexts = res.all_device_contexts;
+    default_device_context = res.default_device_context;
+}
 OCLEngine::OCLEngine(OCLEngine const&) {}
 OCLEngine& OCLEngine::operator=(OCLEngine const& rhs) { return *this; }
 
-void OCLEngine::InitOCL()
+cl::Program OCLEngine::MakeProgram(
+    bool buildFromSource, cl::Program::Sources sources, std::string path, std::shared_ptr<OCLDeviceContext> devCntxt)
+{
+    FILE* clBinFile;
+    cl::Program program;
+    cl_int buildError = -1;
+    std::vector<int> binaryStatus;
+    if (!buildFromSource && (clBinFile = fopen(path.c_str(), "r"))) {
+        long lSize;
+
+        fseek(clBinFile, 0L, SEEK_END);
+        lSize = ftell(clBinFile);
+        rewind(clBinFile);
+
+        std::vector<unsigned char> buffer(lSize);
+        lSize = fread(&buffer[0], sizeof(unsigned char), lSize, clBinFile);
+        fclose(clBinFile);
+
+#if defined(__APPLE__)
+        program = cl::Program(devCntxt->context, { devCntxt->device },
+            { std::pair<const void*, unsigned long>(&buffer[0], buffer.size()) }, &binaryStatus, &buildError);
+#else
+        program = cl::Program(devCntxt->context, { devCntxt->device }, { buffer }, &binaryStatus, &buildError);
+#endif
+
+        if ((buildError != CL_SUCCESS) || (binaryStatus[0] != CL_SUCCESS)) {
+            std::cout << "Binary error: " << buildError << ", " << binaryStatus[0] << " (Falling back to JIT.)"
+                      << std::endl;
+        } else {
+            std::cout << "Loaded binary." << std::endl;
+        }
+    }
+
+    // If, either, there are no cached binaries, or binary loading failed, then fall back to JIT.
+    if (buildError != CL_SUCCESS) {
```

Comment:
Check binaryStatus here as well?

Reply:
Not necessary. `buildError` is only `CL_SUCCESS` if the build was attempted and succeeded.

Reply:
If that's the case, then why do you double-check on ln155?

Reply:
It doesn't hurt to also check on ln155, but both might be unnecessary, though safer. Done.